[RFC] Windows: Quote command invocation when the &shell is not used

equalsraf opened the pull request:
Neovim invokes external commands in two ways, system("cmd arg") calls
cmd through the &shell and applies shell escaping options (just like in Vim).
But system(['cmd', 'arg']) finds cmd in the PATH and executes the command
directly with no shell escaping - both cases are handled by libuv's uv_spawn.

Internaly libuv uses execvp() for UNIX and CreateProcessW() for Windows. The
later takes the arguments as a single string and needs escaping in order to
separate arguments. Libuv does this by default, but we disabled this beaviour
in #3305 to enable system("...") to work as intended, however this is needed
for system([...]).

This commit changes the internal Neovim API (Process) to enable selective
quoting in windows - we want libuv quoting whenever the &shell is not used.
By default quoting is enabled (as was before #3305), but is explicitly set
for system() and jobstart().
- Cherry picked from #810, fixes #3747, unblocks #2773.
- Also added some minor changes to comments and docs - #3518.

PS: admittedly not ideal, this adds a `quote_cmd` parameter to several internal functions that is only used in windows. Maybe there is a better way?

equalsraf committed:
Windows: Quote command invocation when the &shell is not used

Neovim invokes external commands in two ways, system("cmd arg") calls
cmd through the &shell and applies shell escaping options (just like in Vim).
But system(['cmd', 'arg']) finds cmd in the PATH and executes the command
directly with no shell escaping - both cases are handled by libuv's uv_spawn.

Internaly libuv uses execvp() for UNIX and CreateProcessW() for Windows. The
later takes the arguments as a single string and needs escaping in order to
separate arguments. Libuv does this by default, but we disabled this beaviour
in #3305 to enable system("...") to work as intended, however the previous
behaviour is needed for system([...]).

This commit changes the internal Neovim API (Process) to enable selective
quoting in windows - we want libuv quoting whenever the &shell is not used.
By default quoting is enabled (as was before #3305), but is explicitly set
for system() and jobstart().

Added a functional test for system([...]) vs system("...") verifies the
number of arguments received by an external script is the same for both.

Fixes #3747, unblocks #2773, changed docs and comments to address #3518.

splinterofchaos commented on src/nvim/eval.c:
```diff
@@ -11564,16 +11564,30 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
+///
+/// @param cmd_tv If cmd is a string the produced command will include the
+///               &shell, if it is a list it is invoked directly, see the
+///               system() documentation.
```
Please add:
```c
/// @param[out] cmd Set to the string value of the command to execute.
```

splinterofchaos commented:
LGTM :+1: 

Can we add a test demonstrating the problem?

equalsraf commented:
Added missing comment.

> Can we add a test demonstrating the problem?

Added a test, but it might be broken for windows, did not get a chance to test the check yet (in any case tests are not working in windows).

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -162,6 +163,8 @@ int os_call_shell(char_u *cmd, ShellOpts opts, char_u *extra_args)
 /// @param input The input to the shell (NULL for no input), passed to the
```
The comments on `argv`, `input`, and `output` are also outdated (@splinterofchaos please confirm): `do_os_system()` does not internally make any decision about sending a command to the shell--it invokes `argv` directly, which may or may not be a command like `bash -c ...`.

Instead of `shell`, just `process` seems correct.

splinterofchaos replied:
`argv`: Yes. The example is also out of date with respect to that and the new `quote_cmd` argument.

`input`/`output`: Looks correct to me. What might be out of date?

justinmk replied:
input/output both mention "shell" where it should say "process" instead.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -6360,7 +6360,7 @@ synstack({lnum}, {col})					*synstack()*
-		Get the output of the shell command {cmd} as a |string|.  {cmd}
+		Get the output of the command {cmd} as a |string|.  {cmd}
 		will be run the same as in |jobstart()|.  See |systemlist()|
 		to get the output as a |List|.
```
Below this it mentions:

> When {input} is given and is a string this string is written 
> to a file and passed as stdin to the command. 

But `do_os_system()` looks like it sends the input directly to the process.

splinterofchaos replied:
Yup, I didn't update this in #2424 :p 
Oops.

equalsraf replied:
Pushed a couple of commits to update the docs. In the `system()` help I also removed the `Pipes are not used.` line  since that does not seem to be true anymore.

justinmk commented:
@equalsraf Are you aware of https://github.com/neovim/neovim/pull/2789 ?

justinmk commented on test/functional/shell/countargs.bat:
```diff
@@ -0,0 +1,10 @@
+@echo off
```
This file should go in `test/functional/fixtures/system_countargs.bat`

justinmk commented on test/functional/shell/countargs.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
```
This file should go in `test/functional/fixtures/system_countargs.sh`

justinmk commented on test/functional/shell/viml_system_spec.lua:
```diff
@@ -200,6 +201,27 @@ describe('system()', function()
+        fname = "countargs.bat"
+      else
+        fname = "countargs.sh"
+      end
```
local fname = (ffi.os == "Windows" and "countargs.bat" or "countargs.sh")

justinmk commented on test/functional/shell/viml_system_spec.lua:
```diff
@@ -200,6 +201,27 @@ describe('system()', function()
+      if ffi.os == "Windows" then
+        cmd1 = 'system(["cmd", "/c", "' .. script ..'", "1", "2 3"])'
+      else
+        cmd1 = 'system(["' .. script ..'", "1", "2 3"])'
```
To be equivalent with the `cmd /c` invocation above, shouldn't this be:

    cmd1 = 'system(["sh", "-c", "' .. script ..'", "1", "2 3"])'

equalsraf replied:
I only added `cmd /c` because its the only way to run a .bat file :D (CreateProcess does not go throught the shell, unlike exec in UNIX)

justinmk commented on test/functional/shell/viml_system_spec.lua:
```diff
@@ -200,6 +201,27 @@ describe('system()', function()
+        cmd1 = 'system(["' .. script ..'", "1", "2 3"])'
+      end
+      local out1 = eval(cmd1)
+      eq(out0, out1)
```
I'd suggest hard-coding the expected results to avoid hiding a bug that causes both system() variants to return the same wrong output.

```lua
eq("2", out0)
eq("2", out1)
```

justinmk commented on src/nvim/event/process.h:
```diff
@@ -26,6 +26,9 @@ struct process {
   bool closed, term_sent;
+  // Windows only: enclose command arguments in double quotes. This should be
+  // false when using &shell since it has its own escaping
+  bool quote_cmd;
```
`quote_args` ?

equalsraf commented:
> @equalsraf Are you aware of #2789 ?

I am, but #2789 (it was the reason I disabled the quotes in the first place) it needs commands without quotes for compatibility with Vim, while system([...]) needs commands with quotes. This should not break it.

equalsraf commented:
Failing on clint, its strange that it was not failing before this, the line was already folded like that.

ZyX-I commented:
@equalsraf Error suppression suppresses errors in clusters of three lines: line with error, previous line and next line. Should you edit any of the three error is no longer suppressed.

equalsraf commented:
Should I had a NOLINT there?

equalsraf commented:
Ping, rebased against master.

equalsraf commented:
Closing this and reopening later. It seems to me we are doing it wrong here. The right way is to re-enable the quoting from libuv and figure out what is going on with the argv being generated by shell_build_argv instead.
